ast/src/parsed/display.rs

```diff
@@ -564,6 +564,90 @@ fn format_list<L: IntoIterator<Item = I>, I: Display>(list: L) -> String {
     format!("{}", list.into_iter().format(", "))
 }
 
+impl<E: Display> Expression<E> {
+    pub fn precedence(&self) -> Option<ExpressionPrecedence> {
+        match self {
+            Expression::UnaryOperation(op, _) => Some(op.precedence()),
+            Expression::BinaryOperation(_, op, _) => Some(op.precedence()),
+            _ => None,
+        }
+    }
+
+    pub fn require_parentheses(&self) -> bool {
+        match self {
+            Expression::BinaryOperation(_, op, _) => {
+                op.associativity() == BinaryOperatorAssociativity::RequireParentheses
+            }
+            _ => false,
+        }
+    }
+
+    pub fn format_unary_operation(&self, f: &mut Formatter<'_>) -> Result {
+        match self {
+            Expression::UnaryOperation(op, exp) => {
+                let _exp_string = if let (Some(_precision), Some(_exp_precision)) =
+                    (self.precedence(), exp.precedence())
+                {
+                    if _precision < _exp_precision {
+                        format!("({})", exp)
+                    } else {
+                        format!("{}", exp)
+                    }
+                } else {
+                    format!("{}", exp)
+                };
+
+                if op.is_prefix() {
+                    write!(f, "{}{}", op, _exp_string)
+                } else {
+                    write!(f, "{}{}", _exp_string, op)
+                }
+            }
+            _ => panic!("Format with precedence called on non-unary operation."),
+        }
+    }
+
+    pub fn format_binary_operation(&self, f: &mut Formatter<'_>) -> Result {
+        match self {
```

Comment:
If you make this a free function and move the `left, op, right` to the parameters, we can save the match.

Reply:
Same with unary.